File: frontend/src/api/rowdata.ts
/**
 * 表数据 API
 */

import request from '@/utils/request'
import type { ApiResponse } from '@/types/api'

/**
 * 筛选条件
 */
export interface FilterCondition {
  field: string
  operator: string
  value: string
}

/**
 * 分页参数
 */
export interface PaginationParams {
  page: number
  pageSize: number
}

/**
 * 列定义（前端）
 */
export interface ColumnDefFrontend {
  name: string
  type: string
  isPrimary: boolean
  isIndex: boolean
}

/**
 * 行数据结果
 */
export interface RowDataResult {
  columns: ColumnDefFrontend[]
  rows: Record<string, unknown>[]
  total: number
  page: number
  pageSize: number
}

/**
 * 获取表数据
 */
export async function getTableData(
  connectionId: number,
  database: string,
  table: string,
  filters: FilterCondition[] = [],
  pagination: PaginationParams = { page: 1, pageSize: 100 }
): Promise<RowDataResult> {
  const response = await request.post<ApiResponse<RowDataResult>>('/tables/data', {
    connection_id: connectionId,
    database,
    table,
    filters,
    ...pagination
  })
  return response.data
}

/**
 * 更新表行数据
 */
export async function updateTableRow(
  connectionId: number,
  database: string,
  table: string,
  primaryKey: Record<string, unknown>,
  updateData: Record<string, unknown>
): Promise<number> {
  const response = await request.post<ApiResponse<{ affected_rows: number }>>('/tables/data/update', {
    connection_id: connectionId,
    database,
    table,
    primary_key: primaryKey,
    update_data: updateData
  })
  return response.data.affected_rows
}

/**
 * 批量插入数据
 */
export interface BatchInsertRequest {
  connection_id: number
  database: string
  table: string
  data: unknown[][]
  columns: string[]
}

export async function batchInsertData(params: BatchInsertRequest): Promise<{ inserted_rows: number }> {
  const response = await request.post<ApiResponse<{ inserted_rows: number }>>('/tables/data/batch-insert', params)
  return response.data
}

/**
 * 批量更新数据
 */
export interface BatchUpdateRequest {
  connection_id: number
  database: string
  table: string
  filters: FilterCondition[]
  update_data: Record<string, unknown>
}

export async function batchUpdateData(params: BatchUpdateRequest): Promise<{ affected_rows: number }> {
  const response = await request.post<ApiResponse<{ affected_rows: number }>>('/tables/data/batch-update', params)
  return response.data
}

/**
 * 批量删除数据
 */
export interface BatchDeleteRequest {
  connection_id: number
  database: string
  table: string
  filters: FilterCondition[]
}

export async function batchDeleteData(params: BatchDeleteRequest): Promise<{ affected_rows: number }> {
  const response = await request.post<ApiResponse<{ affected_rows: number }>>('/tables/data/batch-delete', params)
  return response.data
}

/**
 * 数据对比请求
 */
export interface CompareDataRequest {
  connection_id: number
  database1: string
  table1: string
  database2?: string
  table2: string
  key_columns: string[]
}

export interface CompareDataResponse {
  only_in_table1: Record<string, unknown>[]
  only_in_table2: Record<string, unknown>[]
  different: Array<{
    key: Record<string, unknown>
    table1_data: Record<string, unknown>
    table2_data: Record<string, unknown>
    differences: string[]
  }>
  same_count: number
}

export async function compareData(params: CompareDataRequest): Promise<CompareDataResponse> {
  const response = await request.post<ApiResponse<CompareDataResponse>>('/tables/data/compare', params)
  return response.data
}

/**
 * 查找替换请求
 */
export interface FindReplaceRequest {
  connection_id: number
  database: string
  table: string
  column: string
  find_value: string
  replace_value: string
  filters?: FilterCondition[]
}

export interface FindReplaceResponse {
  affected_rows: number
  matched_rows: number
}

export async function findReplaceData(params: FindReplaceRequest): Promise<FindReplaceResponse> {
  const response = await request.post<ApiResponse<FindReplaceResponse>>('/tables/data/find-replace', params)
  return response.data
}

/**
 * 从索引信息中提取索引字段
 */
export function extractIndexFields(indexes: Array<{ columns: string[] }>): string[] {
  const fields = new Set<string>()
  indexes.forEach(index => {
    index.columns.forEach(col => fields.add(col))
  })
  return Array.from(fields)
}

File: backend/internal/models/rowdata.go
package models

// GetTableDataRequest 获取表数据请求
type GetTableDataRequest struct {
	ConnectionID int                    `json:"connection_id" binding:"required"`
	Database     string                 `json:"database" binding:"required"`
	Table        string                 `json:"table" binding:"required"`
	Filters      []map[string]interface{} `json:"filters,omitempty"`
	Page         int                    `json:"page,omitempty"`
	PageSize     int                    `json:"pageSize,omitempty"` // 前端使用 pageSize
}

// UpdateTableRowDataRequest 更新表行数据请求
type UpdateTableRowDataRequest struct {
	ConnectionID int                    `json:"connection_id" binding:"required"`
	Database     string                 `json:"database" binding:"required"`
	Table        string                 `json:"table" binding:"required"`
	PrimaryKey   map[string]interface{} `json:"primary_key" binding:"required"`
	UpdateData   map[string]interface{} `json:"update_data" binding:"required"`
}

// BatchInsertRequest 批量插入请求
type BatchInsertRequest struct {
	ConnectionID int           `json:"connection_id" binding:"required"`
	Database     string        `json:"database" binding:"required"`
	Table        string        `json:"table" binding:"required"`
	Data         [][]interface{} `json:"data" binding:"required"`
	Columns      []string      `json:"columns" binding:"required"`
}

// BatchUpdateRequest 批量更新请求
type BatchUpdateRequest struct {
	ConnectionID int                    `json:"connection_id" binding:"required"`
	Database     string                 `json:"database" binding:"required"`
	Table        string                 `json:"table" binding:"required"`
	Filters      []map[string]interface{} `json:"filters" binding:"required"` // 筛选条件
	UpdateData   map[string]interface{} `json:"update_data" binding:"required"`
}

// BatchDeleteRequest 批量删除请求
type BatchDeleteRequest struct {
	ConnectionID int                    `json:"connection_id" binding:"required"`
	Database     string                 `json:"database" binding:"required"`
	Table        string                 `json:"table" binding:"required"`
	Filters      []map[string]interface{} `json:"filters" binding:"required"` // 筛选条件
}

// DeleteByPrimaryKeysRequest 按主键删除请求
type DeleteByPrimaryKeysRequest struct {
	ConnectionID  int                       `json:"connection_id" binding:"required"`
	Database      string                    `json:"database" binding:"required"`
	Table         string                    `json:"table" binding:"required"`
	PrimaryKeys   []map[string]interface{}  `json:"primary_keys" binding:"required"` // 主键值列表，每个元素为 {col1: val1, col2: val2}
}

// CompareDataRequest 数据对比请求
type CompareDataRequest struct {
	ConnectionID int    `json:"connection_id" binding:"required"`
	Database1    string `json:"database1" binding:"required"`
	Table1       string `json:"table1" binding:"required"`
	Database2    string `json:"database2,omitempty"` // 为空则使用 database1
	Table2       string `json:"table2" binding:"required"`
	KeyColumns   []string `json:"key_columns" binding:"required"`
}

// CompareDataResponse 数据对比响应
type CompareDataResponse struct {
	OnlyInTable1 []map[string]interface{} `json:"only_in_table1"`
	OnlyInTable2 []map[string]interface{} `json:"only_in_table2"`
	Different    []struct {
		Key        map[string]interface{} `json:"key"`
		Table1Data map[string]interface{} `json:"table1_data"`
		Table2Data map[string]interface{} `json:"table2_data"`
		Differences []string              `json:"differences"`
	} `json:"different"`
	SameCount int `json:"same_count"`
}

// FindReplaceRequest 查找替换请求
type FindReplaceRequest struct {
	ConnectionID int                    `json:"connection_id" binding:"required"`
	Database     string                 `json:"database" binding:"required"`
	Table        string                 `json:"table" binding:"required"`
	Column       string                 `json:"column" binding:"required"`
	FindValue    string                 `json:"find_value" binding:"required"`
	ReplaceValue string                 `json:"replace_value" binding:"required"`
	Filters      []map[string]interface{} `json:"filters,omitempty"` // 可选的筛选条件
}

// FindReplaceResponse 查找替换响应
type FindReplaceResponse struct {
	AffectedRows int `json:"affected_rows"`
	MatchedRows  int `json:"matched_rows"`
}
